### DISCUSSION TITLE
How to set custom attributes at child spans?

### DISCUSSION THREAD

**Discussion by athoscouto** [open-telemetry/opentelemetry-go#5090] [Q&A] [answered]
I have a common problem where I want to add custom attributes to spans generated by external libraries.
Since the spans are generated inside the library functions that I call, I don't know the proper way to get them and add attributes.

Some attributes can be known ahead of time, but others must be generated when the span is being created (e.g. file and line of the caller function).

Example of instrumentation libraries where I have this problem
- Spans for external HTTP requests: https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/net/http/otelhttp
- Spans for database operations: https://github.com/go-gorm/opentelemetry

This is different from https://github.com/open-telemetry/opentelemetry-go-contrib where my code executes under the framework and I can simply get the span from context and add attributes. 

What would be the proper way to go about it?
Should those libraries provide support for generating those spans?
Or can we do it with the tooling that OTEL provides?

Thanks!

**Comment by dmathieu** [comment]
You can use [SpanFromContext](https://pkg.go.dev/go.opentelemetry.io/otel/trace#SpanFromContext) to retrieve the current span from a `context.Context`.
So if you want to enrich the otelhttp span, in a child handler, you can retrieve the span and add attributes to it.

**Reply by athoscouto** [reply]
@dmathieu I may be missing something obvious here, but I don't think that is possible in my case.
`otelhttp` creates a span inside code run by the `http` package, so I don't generally have a way to get inside that execution and retrieve the current span from the context.

Granted, a workaround would be to create my own transport and mimic what `otelhttp.NewTransport` does:
```go
transport := /* get original transport from somewhere */
transport = transportThatInjectsSpanAttributes(transport)
transport = otelhttp.NewTransport(transport)
```

TBH, I actually like this idea.
I could make it work for the https://github.com/go-gorm/opentelemetry plugin as well by adding another plugin after/before it.

That is way better than my current approach, which involves creating a new HTTP client for every outbound request.

I'll move forward with this approach.
But let me know if I'm missing something and there is some easier way to achieve this.

**Reply by dmathieu** [reply, marked as answer]
Yes, the approach you describe is what I mentioned.
Within your transport, you would have something like this:

```golang
func (t *yourTransport) RoundTrip(r *http.Request) (*http.Response, error) {
  span := trace.SpanFromContext(r.Context())
  span.SetAttributes(attributes.String("myTransport", "is here"))

  return t.parent.RoundTrip()
}
```
